### PR TITLE
Refactor dimension association tables

### PR DESCRIPTION
**Pull request by daniel-thom**
This PR refactor dimension association tables in preparation for registering AEO growth rates.

1. Store association tables by data source.
2. Cache the tables in the spark warehouse so that they don't have to be regenerated at every dataset submission.

I also added a more generic way to register projects and datasets.

### CODE REVIEW

File: dsgrid/utils/spark.py [hunk 1]
```diff
@@ -407,3 +408,102 @@ def _get_spark_session():
...
+    if spark.catalog.tableExists(table_name, dbName=database):
+        return spark.table(table_name)
+    table = try_load_stored_table_from_warehouse(table_name)
+    if table is not None:
+        logger.warning("Table %s was in warehouse but not spark catalog", table_name)
+    return table
+
+
+def is_table_stored(table_name, database="default"):
+    return table_name in list_tables(database=database)
```
Comment (lixiliu) [inline dsgrid/utils/spark.py:437]: I assume spark_db and spark warehouse are two different things and this check whether it's stored in either (as opposed to `_is_table_stored_in_spark_db`) ?

Comment (daniel-thom) [reply dsgrid/utils/spark.py:437]: Based on the documentation you should be able to reuse tables across sessions, like this
```
$ pyspark
In [1]: spark.createDataFrame([{"a": 1}]).write.saveAsTable("my_table")
In [2]: exit()
$ pyspark
In [1]: spark.sql("select * from my_table")
```
It does work for me generally, but not always within dsgrid and I don't know why. I gave up debugging and just started checking in the warehouse directory manually. Refer to line 471

File: dsgrid/utils/spark.py [hunk 2]
```diff
@@ -407,3 +408,102 @@ def _get_spark_session():
...
+    for table in _get_warehouse_path().iterdir():
+        if table.name == table_name:
+            spark = _get_spark_session()
+            return spark.read.parquet(str(table))
+    return None
+
+
+def _delete_table_if_present_but_not_stored(table_name):
+    if not _is_table_stored_in_spark_db(table_name):
+        for table in list(_get_warehouse_path().iterdir()):
```
Comment (lixiliu) [inline dsgrid/utils/spark.py:488]: I am confused. `_is_table_stored_in_spark_db` suggests that spark_db refers to `spark.catalog.listTables(dbName=database)`. But here it seems to be a different repo: `list(_get_warehouse_path().iterdir())`

File: dsgrid/utils/spark.py [hunk 3]
```diff
@@ -407,3 +408,102 @@ def _get_spark_session():
...
+    if not _delete_table_if_present_but_not_stored(table_name) and _is_table_stored_in_spark_db(
+        table_name, database=database
+    ):
+        spark.sql(f"DROP TABLE {table_name}")
+        logger.info("Dropped table %s", table_name)
+
+
+# TODO
+# The next functions are workarounds for a problem where a new SparkSession does not load
+# tables saved in a previous session. Need to debug.
```
Comment (lixiliu) [inline dsgrid/utils/spark.py:473]: This alludes that we may not have a persisted catalog (spark_db, meta_store) within dsgrid

File: tests/data/standard_scenarios_registration.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      "register_project": true,
+      "datasets": [
+        {
+          "dataset_id": "conus_2022_reference_comstock",
+          "dataset_path": "../dsgrid-data/conus_2022_reference_comstock",
+          "config_file": "../dsgrid-project-StandardScenarios/dsgrid_project/datasets/modeled/comstock/dataset.toml",
+          "dimension_mapping_file": "../dsgrid-project-StandardScenarios/dsgrid_project/datasets/modeled/comstock/dimension_mappings.toml",
+          "register_dataset": true,
+          "submit_to_project": true,
+          "fix_dimension_uuids": false,
```
Comment (lixiliu) [inline tests/data/standard_scenarios_registration.json:17]: what does the fix refer to? Replacing the uuids between config files automatically after registration?

Comment (daniel-thom) [reply tests/data/standard_scenarios_registration.json:17]: Suppose a dataset uses a dimension created by the project. The dataset contributor has to put the dimension_id (with UUID) in the file as a dimension reference. If the project gets re-created, that UUID will be wrong. We have code that will update the UUIDs.

File: dsgrid/config/project_config.py [hunk 5]
```diff
@@ -730,43 +732,142 @@ def _iter_datasets_by_status(self, status: DatasetRegistryStatus) -> InputDatase
...
+            logger.info("Start processing dimension=%s table_count=%s", dim, table_count)
             if table is None:
                 table = other
             else:
                 table = table.crossJoin(other)
-                assert table.count() == other.count() * table_count
+                count = table.count()
+                assert count == other.count() * table_count
+                table_count = count
+            other.unpersist()
```
Comment (lixiliu) [inline dsgrid/config/project_config.py:811]: What about unpersisting `other_dims[dim]`?

File: dsgrid/config/project_config.py [hunk 6]
```diff
@@ -730,43 +732,142 @@ def _iter_datasets_by_status(self, status: DatasetRegistryStatus) -> InputDatase
...
+                if found:
+                    raise DSGInvalidDimensionAssociation(
+                        f"dimension association for {dim_type} has supplemental record ID "
+                        "{supplemental_id} defined in at least two supplemental dimensions"
+                    )
+                found = True
+                new_records = (
+                    table.join(
+                        mapping_records,
+                        on=table[dim_type.value] == mapping_records.to_id,
```
Comment (lixiliu) [inline dsgrid/config/project_config.py:855]: I think this is a one-to-many join, with one value `table[dim_type.value]` mappable to many values in `mapping_records.to_id`. Did the inner join produce  join as expected? And do you need to prefilter `mapping_records` before the inner join here?

Comment (lixiliu) [reply dsgrid/config/project_config.py:855]: What happens if you do table.left_join(mapping_records)? Could you get the new table directly without doing the union below?

**Review by lixiliu — COMMENTED** [review]
okay, I'll stop now

**Pull request closed without merging**